### ISSUE TITLE
ICE from const item in lifetime-parametric impl

### ISSUE THREAD

**Issue by thejpster** [rust-lang/rust#56445]
Reduced test case ([play](https://play.rust-lang.org/?version=stable&mode=debug&edition=2015&gist=d24bad9fcc0009f035af607f2dd90522)) from [comment below](https://github.com/rust-lang/rust/issues/56445#issuecomment-502095133):

```rust
struct OnDiskDirEntry<'a> { _s: &'a usize }

impl<'a> OnDiskDirEntry<'a> {
    const LFN_FRAGMENT_LEN: usize = 2;

    fn lfn_contents(&self) -> [char; Self::LFN_FRAGMENT_LEN] { loop { } }
}
```

Another repro case: https://github.com/rust-lang/rust/issues/56445#issuecomment-524494170

Original bug report follows

----

If I run `cargo test` on https://github.com/thejpster/embedded-sdmmc-rs/tree/generates_ice, I get an ICE:

```
user@host:~/Documents/programming/embedded-sdmmc$ cargo test
   Compiling embedded-sdmmc v0.1.0 (/home/jonathan/Documents/programming/embedded-sdmmc)                                                                                                                    
thread 'main' panicked at 'assertion failed: `(left == right)`                                                                                                                                              
  left: `Some(NodeId(1093))`,                                                                                                                                                                               
 right: `None`: free_scope: DefId(0/0:72 ~ embedded_sdmmc[4f8a]::fat[0]::{{impl}}[4]) not recognized by the region scope tree for None / Some(DefId(0/0:79 ~ embedded_sdmmc[4f8a]::fat[0]::{{impl}}[4]::lfn_contents[0]))', librustc/middle/region.rs:673:13
note: Run with `RUST_BACKTRACE=1` for a backtrace.                                                                                                                                                          
                                                                                                                                                                                                            
error: internal compiler error: unexpected panic                                                                                                                                                            
                                                                                                                                                                                                            
note: the compiler unexpectedly panicked. this is a bug.                                                                                                                                                    
                                                                                                                                                                                                            
note: we would appreciate a bug report: https://github.com/rust-lang/rust/blob/master/CONTRIBUTING.md#bug-reports                                                                                           
                                                                                                                                                                                                            
note: rustc 1.31.0-beta.19 (42053f9f0 2018-11-26) running on x86_64-unknown-linux-gnu                                                                                                                       
                                                                                                                                                                                                            
note: compiler flags: -C debuginfo=2 -C incremental --crate-type lib                                                                                                                                        
                                                                                                                                                                                                            
note: some of the compiler flags provided by cargo are hidden                                                                                                                                               
                                                                                                                                                                                                            
error: Could not compile `embedded-sdmmc`.                                                                                                                                                                  
warning: build failed, waiting for other jobs to finish...
thread 'main' panicked at 'assertion failed: `(left == right)`                                                                                                                                              
  left: `Some(NodeId(1090))`,                                                                                                                                                                               
 right: `None`: free_scope: DefId(0/0:71 ~ embedded_sdmmc[be94]::fat[0]::{{impl}}[4]) not recognized by the region scope tree for None / Some(DefId(0/0:78 ~ embedded_sdmmc[be94]::fat[0]::{{impl}}[4]::lfn_contents[0]))', librustc/middle/region.rs:673:13
note: Run with `RUST_BACKTRACE=1` for a backtrace.                                                                                                                                                          
                                                                                                                                                                                                            
error: internal compiler error: unexpected panic                                                                                                                                                            
                                                                                                                                                                                                            
note: the compiler unexpectedly panicked. this is a bug.                                                                                                                                                    
                                                                                                                                                                                                            
note: we would appreciate a bug report: https://github.com/rust-lang/rust/blob/master/CONTRIBUTING.md#bug-reports                                                                                           
                                                                                                                                                                                                            
note: rustc 1.31.0-beta.19 (42053f9f0 2018-11-26) running on x86_64-unknown-linux-gnu                                                                                                                       
                                                                                                                                                                                                            
note: compiler flags: -C debuginfo=2 -C incremental                                                                                                                                                         
                                                                                                                                                                                                            
note: some of the compiler flags provided by cargo are hidden                                                                                                                                               
                                                                                                                                                                                                            
error: Could not compile `embedded-sdmmc`.                                                                                                                                                                  

To learn more, run the command again with --verbose.

user@host:~/Documents/programming/embedded-sdmmc$ cargo +nightly test
   Compiling embedded-sdmmc v0.1.0 (/home/jonathan/Documents/programming/embedded-sdmmc)
thread 'main' panicked at 'assertion failed: `(left == right)`
  left: `Some(NodeId(1093))`,
 right: `None`: free_scope: DefId(0/0:72 ~ embedded_sdmmc[46a7]::fat[0]::{{impl}}[4]) not recognized by the region scope tree for None / Some(DefId(0/0:79 ~ embedded_sdmmc[46a7]::fat[0]::{{impl}}[4]::lfn_contents[0]))', src/librustc/middle/region.rs:673:13
note: Run with `RUST_BACKTRACE=1` for a backtrace.

error: internal compiler error: unexpected panic

note: the compiler unexpectedly panicked. this is a bug.

note: we would appreciate a bug report: https://github.com/rust-lang/rust/blob/master/CONTRIBUTING.md#bug-reports

note: rustc 1.32.0-nightly (400c2bc5e 2018-11-27) running on x86_64-unknown-linux-gnu

note: compiler flags: -C debuginfo=2 -C incremental --crate-type lib

note: some of the compiler flags provided by cargo are hidden

error: Could not compile `embedded-sdmmc`.
warning: build failed, waiting for other jobs to finish...
thread 'main' panicked at 'assertion failed: `(left == right)`
  left: `Some(NodeId(1090))`,
 right: `None`: free_scope: DefId(0/0:71 ~ embedded_sdmmc[93d1]::fat[0]::{{impl}}[4]) not recognized by the region scope tree for None / Some(DefId(0/0:78 ~ embedded_sdmmc[93d1]::fat[0]::{{impl}}[4]::lfn_contents[0]))', src/librustc/middle/region.rs:673:13
note: Run with `RUST_BACKTRACE=1` for a backtrace.

error: internal compiler error: unexpected panic

note: the compiler unexpectedly panicked. this is a bug.

note: we would appreciate a bug report: https://github.com/rust-lang/rust/blob/master/CONTRIBUTING.md#bug-reports

note: rustc 1.32.0-nightly (400c2bc5e 2018-11-27) running on x86_64-unknown-linux-gnu

note: compiler flags: -C debuginfo=2 -C incremental

note: some of the compiler flags provided by cargo are hidden

error: Could not compile `embedded-sdmmc`.

To learn more, run the command again with --verbose.

user@host:~/Documents/programming/embedded-sdmmc$ cargo version
cargo 1.31.0-beta (339d9f9c8 2018-11-16)

user@host:~/Documents/programming/embedded-sdmmc$ cargo +nightly version
cargo 1.32.0-nightly (b3d0b2e54 2018-11-15)
```

I'm using 2018 edition syntax, so I can only test on beta and nightly.

<!-- TRIAGEBOT_START -->

<!-- TRIAGEBOT_ASSIGN_START -->

<!-- TRIAGEBOT_ASSIGN_DATA_START$${"user":"sjakobi"}$$TRIAGEBOT_ASSIGN_DATA_END -->

<!-- TRIAGEBOT_ASSIGN_END -->
<!-- TRIAGEBOT_END -->

**Comment by thejpster** [comment]
Commenting out the `lfn_contents()` function in fat.rs makes the ICE go away.

The ICE occurs on a `cargo build` as well as a `cargo test`.

**Comment by thejpster** [comment]
Replacing both instances of Self::FRAGMENT_LEN in the `lfn_contents()` function with the literal integer 13 makes the ICE go away.

**Comment by jonas-schievink** [comment]
Still reproduces on 1.35.0. Backtrace:

```
stack backtrace:
   0: std::sys::unix::backtrace::tracing::imp::unwind_backtrace
             at src/libstd/sys/unix/backtrace/tracing/gcc_s.rs:39
   1: std::sys_common::backtrace::_print
             at src/libstd/sys_common/backtrace.rs:71
   2: std::panicking::default_hook::{{closure}}
             at src/libstd/sys_common/backtrace.rs:59
             at src/libstd/panicking.rs:197
   3: std::panicking::default_hook
             at src/libstd/panicking.rs:211
   4: rustc::util::common::panic_hook
   5: std::panicking::rust_panic_with_hook
             at src/libstd/panicking.rs:478
   6: std::panicking::continue_panic_fmt
             at src/libstd/panicking.rs:381
   7: std::panicking::begin_panic_fmt
             at src/libstd/panicking.rs:336
   8: rustc::middle::region::ScopeTree::early_free_scope
   9: rustc::middle::free_region::RegionRelations::is_subregion_of
  10: rustc::infer::lexical_region_resolve::resolve
  11: rustc::infer::InferCtxt::resolve_regions_and_report_errors
  12: rustc_typeck::check::regionck::<impl rustc_typeck::check::FnCtxt>::regionck_expr
  13: rustc::ty::context::GlobalCtxt::enter_local
  14: rustc_typeck::check::typeck_tables_of
  15: rustc::ty::query::__query_compute::typeck_tables_of
  16: rustc::ty::query::<impl rustc::ty::query::config::QueryAccessors for rustc::ty::query::queries::typeck_tables_of>::compute
  17: rustc::dep_graph::graph::DepGraph::with_task_impl
  18: rustc::ty::query::plumbing::<impl rustc::ty::context::TyCtxt>::get_query
  19: rustc_mir::const_eval::const_eval_raw_provider
  20: rustc::ty::query::__query_compute::const_eval_raw
  21: rustc::ty::query::<impl rustc::ty::query::config::QueryAccessors for rustc::ty::query::queries::const_eval_raw>::compute
  22: rustc::dep_graph::graph::DepGraph::with_task_impl
  23: rustc::ty::query::plumbing::<impl rustc::ty::context::TyCtxt>::get_query
  24: rustc_mir::const_eval::const_eval_provider
  25: rustc::ty::query::__query_compute::const_eval
  26: rustc::ty::query::<impl rustc::ty::query::config::QueryAccessors for rustc::ty::query::queries::const_eval>::compute
  27: rustc::dep_graph::graph::DepGraph::with_task_impl
  28: rustc::ty::query::plumbing::<impl rustc::ty::context::TyCtxt>::get_query
  29: <rustc::traits::project::AssociatedTypeNormalizer as rustc::ty::fold::TypeFolder>::fold_const
  30: rustc::ty::structural_impls::<impl rustc::ty::fold::TypeFoldable for &rustc::ty::TyS>::super_fold_with
  31: <rustc::traits::project::AssociatedTypeNormalizer as rustc::ty::fold::TypeFolder>::fold_ty
  32: <smallvec::SmallVec<A> as core::iter::traits::collect::FromIterator<<A as smallvec::Array>::Item>>::from_iter
  33: rustc::ty::fold::TypeFoldable::fold_with
  34: rustc::ty::structural_impls::<impl rustc::ty::fold::TypeFoldable for &rustc::ty::TyS>::super_fold_with
  35: <rustc::traits::project::AssociatedTypeNormalizer as rustc::ty::fold::TypeFolder>::fold_ty
  36: <smallvec::SmallVec<A> as core::iter::traits::collect::FromIterator<<A as smallvec::Array>::Item>>::from_iter
  37: rustc::ty::fold::TypeFoldable::fold_with
  38: rustc::ty::structural_impls::<impl rustc::ty::fold::TypeFoldable for &rustc::ty::TyS>::super_fold_with
  39: <rustc::traits::project::AssociatedTypeNormalizer as rustc::ty::fold::TypeFolder>::fold_ty
  40: <smallvec::SmallVec<A> as core::iter::traits::collect::FromIterator<<A as smallvec::Array>::Item>>::from_iter
  41: rustc::ty::fold::TypeFoldable::fold_with
  42: rustc::traits::project::normalize
  43: rustc::infer::InferCtxt::partially_normalize_associated_types_in
  44: rustc::ty::context::GlobalCtxt::enter_local
  45: rustc_typeck::check::wfcheck::check_associated_item
  46: rustc::ty::query::__query_compute::check_impl_item_well_formed
  47: rustc::ty::query::<impl rustc::ty::query::config::QueryAccessors for rustc::ty::query::queries::check_impl_item_well_formed>::compute
  48: rustc::dep_graph::graph::DepGraph::with_task_impl
  49: rustc::ty::query::plumbing::<impl rustc::ty::context::TyCtxt>::get_query
  50: <rustc_typeck::check::wfcheck::CheckTypeWellFormedVisitor as rustc::hir::itemlikevisit::ParItemLikeVisitor>::visit_impl_item
  51: __rust_maybe_catch_panic
             at src/libpanic_unwind/lib.rs:87
  52: rustc_data_structures::sync::par_for_each_in
  53: __rust_maybe_catch_panic
             at src/libpanic_unwind/lib.rs:87
  54: rustc::hir::Crate::par_visit_all_item_likes
  55: rustc::util::common::time
  56: rustc_typeck::check_crate
  57: rustc_interface::passes::analysis
  58: rustc::ty::query::__query_compute::analysis
  59: rustc::ty::query::<impl rustc::ty::query::config::QueryAccessors for rustc::ty::query::queries::analysis>::compute
  60: rustc::dep_graph::graph::DepGraph::with_task_impl
  61: rustc::ty::query::plumbing::<impl rustc::ty::context::TyCtxt>::get_query
  62: rustc_interface::passes::BoxedGlobalCtxt::access::{{closure}}
  63: rustc_interface::passes::create_global_ctxt::{{closure}}
  64: rustc_interface::interface::run_compiler_in_existing_thread_pool
  65: std::thread::local::LocalKey<T>::with
  66: scoped_tls::ScopedKey<T>::set
  67: syntax::with_globals
query stack during panic:
#0 [typeck_tables_of] processing `fat::OnDiskDirEntry::<'a>::lfn_contents::{{constant}}#0`
#1 [const_eval_raw] const-evaluating `fat::OnDiskDirEntry::<'a>::lfn_contents::{{constant}}#0`
#2 [const_eval] const-evaluating + checking `fat::OnDiskDirEntry::<'a>::lfn_contents::{{constant}}#0`
#3 [check_impl_item_well_formed] processing `fat::OnDiskDirEntry::<'a>::lfn_contents`
#4 [analysis] running analysis passes on this crate
end of query stack
```

**Comment by pnkfelix** [comment]
triage: P-high, removing nomination, assigning to self for initial investigation.

**Comment by pnkfelix** [comment]
Note: I see a different ICE on nightly on this branch, namely:
```
error: internal compiler error: src/librustc_mir/borrow_check/nll/universal_regions.rs:741: cannot convert `ReEarlyBound(0, 'a)` to a region vid
```

**Comment by pnkfelix** [comment]
Okay here is reduced test case:

```rust
struct OnDiskDirEntry<'a> {}

impl<'a> OnDiskDirEntry<'a> {
    const LFN_FRAGMENT_LEN: usize = 13;

    fn lfn_contents(&self) -> Option<(bool, u8, [char; Self::LFN_FRAGMENT_LEN])> { loop { }    }
}
```

**Comment by pnkfelix** [comment]
@thejpster another workaround, by the way, is to move the `const LFN_FRAGMENT_LEN` out of the `impl` (and then change the use to say `LFN_FRAGMENT_LEN` instead of `Self::LFN_FRAGMENT_LEN`)

**Comment by pnkfelix** [comment]
And yet another workaround is to say `[char; OnDiskDirEntry::LFN_FRAGMENT_LEN]` instead of `[char; Self::LFN_FRAGMENT_LEN]` in the return type.

**Comment by pnkfelix** [comment]
I am taking a break for a few months and I wont be able to look at this; unassigning self.

**Comment by estebank** [comment]
[Another repro case taken](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=b3d5aa637469eb450d27d73f35ac12f4) from duplicate #63290:

```
#![feature(const_generics)]
use core::marker::PhantomData;

struct Bug<'a, const S: &'a str>(PhantomData<&'a ()>);

impl Bug<'_, {""}> {}
```

**Comment by jonas-schievink** [comment]
Another occurrence in https://github.com/rust-lang/rust/issues/63838:

```rust
pub struct Memory<'rom> {
    rom: &'rom [u8],
    ram: [u8; Self::SIZE],
}

impl<'rom> Memory<'rom> {
    pub const SIZE: usize = 0x8000;
}
```

**Comment by PlasmaPower** [comment]
I've been trying to trace this down, and I think it's due to #43408 (or at least it's the same code that needs fixed there, in rustc_typeck::collect::generics_of). I believe the UniversalRegionIndices aren't being properly populated for the constant, because in the HIR the constant doesn't have a parent. ConstraintConversion needs the UniversalRegionIndices to convert the ReEarlyBound into a RegionVid, but it's missing due to the previous problem.

Simply replacing [this line](https://github.com/rust-lang/rust/blob/master/src/librustc_typeck/collect.rs#L910) with `if true {` makes the last Memory examples compile, but the original example fails later in free_scope. I'm super new to the rust compiler, but hopefully someone else can use this information to fix the problem :)

I also believe that the const generics issue is separate, but it's possible that I'm simply trying to solve this in the wrong manner, and a correct solution would also solve the const generics issue.

**Comment by pnkfelix** [comment]
triage: It would be good to at least figure out a way to give a better error message in these array length cases. (other cases include #43408, #52808...)

**Comment by jonas-schievink** [comment]
Another case in https://github.com/rust-lang/rust/issues/72242:

```rust
use std::marker::PhantomData;

struct S<'a>
{
    m1: PhantomData<&'a u8>,
    m2: [u8; S::size()],
}

impl<'a> S<'a>
{
    const fn size() -> usize { 1 }

    fn new() -> Self
    {
        Self
        {
            m1: PhantomData,
            m2: [0; Self::size()],
        }
    }
}
```

**Comment by JohnTitor** [comment]
Triage: [This ICE](https://github.com/rust-lang/rust/issues/56445#issuecomment-629426939) is now fixed in the latest nightly (rustc 1.45.0-nightly (46e85b432 2020-05-24)) (fixed by #72362).

**Comment by JohnTitor** [comment]
Triage: https://github.com/rust-lang/rust/issues/56445#issuecomment-518402995 is now fixed in the latest nightly (I think it's fixed by #74051).
(And I've hidden comment as resolved to make the un-fixed ICEs easier to find.)

**Comment by JohnTitor** [comment]
Triage:
The reduced snippet (https://github.com/rust-lang/rust/issues/56445#issuecomment-502095133) and https://github.com/rust-lang/rust/issues/56445#issuecomment-524494170 are no longer ICE with the latest nightly: https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=2d74015d14721b961d717ac11c343a0c
